code/datums/mood.dm

```diff
@@ -0,0 +1,471 @@
+#define MINOR_INSANITY_PEN 5
+#define MAJOR_INSANITY_PEN 10
+#define MOOD_CATEGORY_NUTRITION "nutrition"
+#define MOOD_CATEGORY_AREA_BEAUTY "area_beauty"
+
+/**
+ * Mood datum
+ *
+ * Contains the logic for controlling a living mob's mood and sanity.
+ */
+/datum/mood
+	/// Weakref to the parent (living) mob
+	var/datum/weakref/parent
+
+	/// Mob's mood
+	var/mood
+	/// Mob's sanity
+	var/sanity = SANITY_NEUTRAL
+	/// The displayed mood
+	var/shown_mood
+	/// Modifier to allow certain mobs to be less affected by moodlets
+	var/mood_modifier = 1
+	/// Used to track what stage of moodies they're on
+	var/mood_level = 5
+	/// To track what stage of sanity they're on
+	var/sanity_level = SANITY_LEVEL_NEUTRAL
+	/// Is the owner being punished for low mood? if so, how much?
+	var/insanity_effect = 0
+	/// The screen object for the current mood level
+	var/atom/movable/screen/mood/mood_screen_object
+
+	/// List of mood events currently active on this datum
+	var/list/mood_events = list()
+
+/datum/mood/New(mob/living/mob_to_make_moody)
+	if (!istype(mob_to_make_moody))
+		stack_trace("Tried to apply mood to a non-living atom!")
+		qdel(src)
+		return
+
+	START_PROCESSING(SSmood, src)
+
+	parent = WEAKREF(mob_to_make_moody)
+
+	RegisterSignal(mob_to_make_moody, COMSIG_MOB_HUD_CREATED, .proc/modify_hud)
+	RegisterSignal(mob_to_make_moody, COMSIG_ENTER_AREA, .proc/check_area_mood)
+	RegisterSignal(mob_to_make_moody, COMSIG_LIVING_REVIVE, .proc/on_revive)
+
+	mob_to_make_moody.become_area_sensitive(MOOD_DATUM_TRAIT)
+	if(mob_to_make_moody.hud_used)
+		modify_hud()
+		var/datum/hud/hud = mob_to_make_moody.hud_used
+		hud.show_hud(hud.hud_version)
+
+/datum/mood/Destroy(force, ...)
+	STOP_PROCESSING(SSmood, src)
+
+	QDEL_LIST_ASSOC_VAL(mood_events)
+
+	unmodify_hud()
+	var/mob/living/mob_parent = parent?.resolve()
+	if (mob_parent)
+		mob_parent.lose_area_sensitivity(MOOD_DATUM_TRAIT)
+		UnregisterSignal(mob_parent, list(COMSIG_MOB_HUD_CREATED, COMSIG_ENTER_AREA, COMSIG_LIVING_REVIVE))
+
+	return ..()
+
+/datum/mood/process(delta_time)
+	var/mob/living/mob_parent = parent?.resolve()
+	if (mob_parent.stat == DEAD)
```

Comment:
could you make it stop processing if they die and resume if they undie (you can register to COMSIG_MOB_STATCHANGE for this, no stat changes happen outside of set_stat())

Reply:
sure